compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs

```diff
@@ -212,128 +222,120 @@ fn save_function_record(
     for &non_codegenned_def_id in all_def_ids.difference(codegenned_def_ids) {
-        // Make sure the non-codegenned (unreachable) function has a file_name
+        // Make sure the non-codegenned (unused) function has a file_name
         if let Some(non_codegenned_file_name) = tcx.covered_file_name(non_codegenned_def_id) {
-            let def_ids = unreachable_def_ids_by_file
-                .entry(*non_codegenned_file_name)
-                .or_insert_with(Vec::new);
+            let def_ids =
+                unused_def_ids_by_file.entry(*non_codegenned_file_name).or_insert_with(Vec::new);
             def_ids.push(non_codegenned_def_id);
         }
     }
 
-    if unreachable_def_ids_by_file.is_empty() {
-        // There are no unreachable functions with file names to add (in any CGU)
+    if unused_def_ids_by_file.is_empty() {
+        // There are no unused functions with file names to add (in any CGU)
         return;
     }
 
-    // Since there may be multiple `CodegenUnit`s, some codegenned_def_ids may be codegenned in a
-    // different CGU, and will be added to the function_coverage_map for each CGU. Determine which
-    // function_coverage_map has the responsibility for publishing unreachable coverage
-    // based on file name:
+    // Each `CodegenUnit` (CGU) has its own function_coverage_map, and generates a specific binary
+    // with its own coverage map.
+    //
+    // Each covered function `Instance` can be included in only one coverage map, produced from a
+    // specific function_coverage_map, from a specific CGU.
     //
-    // For each covered file name, sort ONLY the non-generic codegenned_def_ids, and if
-    // covered_def_ids.contains(the first def_id) for a given file_name, add the unreachable code
-    // region in this function_coverage_map. Otherwise, ignore it and assume another CGU's
-    // function_coverage_map will be adding it (because it will be first for one, and only one,
-    // of them).
+    // Since unused functions did not generate code, they are not associated with any CGU yet.
+    //
+    // To avoid injecting the unused functions in multiple coverage maps (for multiple CGUs)
+    // determine which function_coverage_map has the responsibility for publishing unreachable
+    // coverage, based on file name: For each unused function, find the CGU that generates the
+    // first function (based on sorted `DefId`) from the same file.
+    //
+    // Add a new `FunctionCoverage` to the `function_coverage_map`, with unreachable code regions
+    // for each region in it's MIR.
+
+    // Convert the `HashSet` of `codegenned_def_ids` to a sortable vector, and sort them.
     let mut sorted_codegenned_def_ids: Vec<DefId> =
         codegenned_def_ids.iter().map(|def_id| *def_id).collect();
     sorted_codegenned_def_ids.sort_unstable();
 
     let mut first_covered_def_id_by_file: FxHashMap<Symbol, DefId> = FxHashMap::default();
```

Comment:
Looking at this with fresh eyes (maybe we discussed this before).. This will only work if there is a covered def in each file containing an uncovered def.

I'm wondering if we can do it a different way by sorting the `Symbol` (which is just an index IIRC), partitioning by the number of codegen units, and using the index of the current codegen unit to determine which files to generate uncovered symbols for. In fact this could use the DefIds directly instead of Symbols.

Of course, that assumes there's a way to get an index of the current codegen unit and the number of codegen units.. not sure if there is or if we can add one.

Reply:
Yeah, it's better than it used to be. Previously the file had to have at least one ***non-generic*** instance. Now it's fine to have only generics.

I thought about this a lot this week. Yes, I agree, I'd love to be able to do this differently, but I don't understand enough about the CGU structure to visualize a better solution.

Can we keep this good, but challenging (for me), suggestion for a separate PR?

Reply:
FYI, as an experiment, I added a library crate called `unused_crate.rs`, and added one `pub const`, then referenced only the const in the `uses_crate.rs` test. If the `unused_crate` defines any public non-generic functions, at least one of them will be eagerly generated into an Instance, and that's enough to have coverage for all of them (even private or generic). But if the crate ONLY has private non-generic and/or public generics, nothing gets coverage. So that's the only case.

HOWEVER! Even when they do get coverage in the coverage map (because there is one public non-generic), `llvm-cov show` STILL doesn't generate a coverage report for that file for whatever reason.

Maybe that's a bug, or just the way it works, but it looks like even if we change the way mapgen.rs works, it will apparently have no net effect or improvement.

Reply:
One clarification, if any function in the crate actually _is_ used, then you should get coverage (including in reports). This is only talking about a create that has no used functions.